pkg/apis/messaging/v1alpha1/pipeline_types.go

```diff
@@ -0,0 +1,154 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/pkg/apis"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	duckv1beta1 "github.com/knative/pkg/apis/duck/v1beta1"
+	"github.com/knative/pkg/webhook"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// Pipeline defines a sequence of Subscribers that will be wired in
+// series through Channels and Subscriptions.
+type Pipeline struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Pipeline.
+	Spec PipelineSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Pipeline. This data may be out of
+	// date.
+	// +optional
+	Status PipelineStatus `json:"status,omitempty"`
+}
+
+// Check that Pipeline can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Pipeline)(nil)
+var _ apis.Defaultable = (*Pipeline)(nil)
+
+// TODO: make appropriate fields immutable.
+//var _ apis.Immutable = (*Pipeline)(nil)
+var _ runtime.Object = (*Pipeline)(nil)
+var _ webhook.GenericCRD = (*Pipeline)(nil)
+
+// This should be duck so that Broker can also use this
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ChannelTemplateSpec struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Spec defines the Spec to use for each channel created. Passed
+	// in verbatim to the Channel CRD as Spec section.
+	// +optional
+	Spec runtime.RawExtension `json:"spec"`
+}
+
+// Internal version of ChannelTemplateSpec that includes ObjectMeta so that
+// we can easily create new Channels off of it.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ChannelTemplateSpecInternal struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the Spec to use for each channel created. Passed
+	// in verbatim to the Channel CRD as Spec section.
+	// +optional
+	Spec runtime.RawExtension `json:"spec"`
+}
+
+type PipelineSpec struct {
+	// Steps is the list of Subscribers (processors / functions) that will be called in the order
+	// provided.
+	Steps []eventingv1alpha1.SubscriberSpec `json:"steps"`
+
+	// ChannelTemplate specifies which Channel CRD to use
+	ChannelTemplate ChannelTemplateSpec `json:"channelTemplate"`
+
+	// Reply is a Reference to where the result of the last Subscriber gets sent to.
+	// Currently due to limitations on the Subscription, this _must_ be a Channel.
```

Comment:
I think this is now out of date. @nachocano, I think you removed this requirement, right?